File: app/src/main/java/bible/translationtools/converter/TransformerFragment.java
package bible.translationtools.converter;

import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;
import bible.translationtools.converterlib.ITransformer;
import bible.translationtools.converterlib.Project;
import bible.translationtools.converterlib.Transformer;
import com.toptoche.searchablespinnerlibrary.SearchableSpinner;

public class TransformerFragment extends Fragment implements TransformerTask.TransformerResultCallback {

    private MainActivity activity;

    private String projectName;
    private String projectsName; // language and version
    private Language language;
    private Version version;
    private String book;
    private boolean isTransforming = false;
    private String messageText = "";
    private String buttonText = "";

    private Language newLanguage;
    private Version newVersion;

    private Button button;
    private ProgressBar progress;
    private TextView projectTitle;
    private TextView messageView;
    private Spinner versionSpinner;
    private SearchableSpinner languageSpinner;
    private CheckBox allBooksCheckbox;

    private LanguageRepository langRepo;
    private ArrayAdapter<Version> versionsAdapter;
    private ArrayAdapter<Language> languageAdapter;

    protected ITransformer transformer;

    public static TransformerFragment newInstance(Project project) {
        Bundle bundle = new Bundle();
        bundle.putString("projectName", project.toString());
        bundle.putString("language", project.language);
        bundle.putString("version", project.version);
        bundle.putString("book", project.book);

        TransformerFragment fragment = new TransformerFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    private void readBundle(Bundle bundle) {
        if (bundle != null) {
            projectName = bundle.getString("projectName");
            projectsName = String.format("%s | %s", bundle.getString("language"), bundle.getString("version"));

            Language lang = langRepo.getLang(bundle.getString("language"));
            language = new Language(
                    lang.slug,
                    lang.name,
                    lang.angName
            );

            Version ver = VersionRepository.getVersion(bundle.getString("version"));
            version = new Version(
                    ver.slug,
                    ver.name
            );
            book = bundle.getString("book");
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        langRepo = new LanguageRepository(getContext());

        readBundle(getArguments());

        return inflater.inflate(R.layout.transformer_fragment, container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        activity = (MainActivity)getActivity();

        button = view.findViewById(R.id.transform);
        progress = view.findViewById(R.id.progressBar);
        projectTitle = view.findViewById(R.id.titleView);
        languageSpinner = view.findViewById(R.id.languages);
        versionSpinner = view.findViewById(R.id.versions);
        messageView = view.findViewById(R.id.messageView);
        allBooksCheckbox = view.findViewById(R.id.allBooksCheckbox);

        buttonText = getString(R.string.transform); // Default value
        languageSpinner.setTitle(getString(R.string.select_language));

        init();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }

    @Override
    public Integer startTransformation() {
        return transformer.execute();
    }

    protected void transform() {
        String dir = activity.bttrDir().getAbsolutePath();
        newLanguage = (Language) languageSpinner.getSelectedItem();
        newVersion = (Version) versionSpinner.getSelectedItem();

        newLanguage = !newLanguage.equals(language) ? newLanguage : null;
        newVersion = !newVersion.equals(version) ? newVersion : null;

        String newLanguageSlug = newLanguage != null ? newLanguage.slug : null;
        String newVersionSlug = newVersion != null ? newVersion.slug : null;

        boolean transformAll = allBooksCheckbox.isChecked();

        try {
            transformer = new Transformer(
                    dir,
                    language.slug,
                    version.slug,
                    (!transformAll ? book : null),
                    newLanguageSlug,
                    null,
                    newVersionSlug
            );
            TransformerTask transformerTask = new TransformerTask(TransformerFragment.this);
            transformerTask.execute();
        } catch (Exception e) {
            Toast.makeText(getActivity().getApplicationContext(), e.getMessage(), Toast.LENGTH_LONG).show();
            e.printStackTrace();
        }
    }

    @Override
    public Void transformationStarted() {
        isTransforming = true;
        messageText = "";
        buttonText = getString(R.string.processing);

        button.setEnabled(false);
        button.setText(buttonText);
        languageSpinner.setVisibility(View.GONE);
        versionSpinner.setVisibility(View.GONE);
        allBooksCheckbox.setVisibility(View.GONE);
        progress.setVisibility(View.VISIBLE);
        messageView.setText(messageText);
        return null;
    }

    @Override
    public Void transformationDone(final Integer result) {
        isTransforming = false;
        buttonText = getString(R.string.transform);
        transformer.setDateTimeDir();
        button.setEnabled(true);
        button.setText(buttonText);
        progress.setVisibility(View.GONE);
        languageSpinner.setVisibility(View.VISIBLE);
        versionSpinner.setVisibility(View.VISIBLE);

        if(result >= 0) {
            messageText = getString(R.string.transformation_complete, result);
            messageView.setText(messageText);

            language = newLanguage != null ? newLanguage : language;
            version = newVersion != null ? newVersion : version;
            projectName = String.format("%s | %s | %s", language.slug, version.slug, book);
            projectsName = String.format("%s | %s", language.slug, version.slug);

            projectTitle.setText(projectName);
            languageSpinner.setSelection(languageAdapter.getPosition(language));
            versionSpinner.setSelection(versionsAdapter.getPosition(version));
            allBooksCheckbox.setText(getString(R.string.all_books_check, projectsName));

            getArguments().putString("projectName", projectName);
            getArguments().putString("language", language.slug);
            getArguments().putString("version", version.slug);
        } else {
            messageText = getString(R.string.error_occurred);
            messageView.setText(messageText);
        }
        messageView.setTextColor(Color.BLACK);
        System.out.println("Finished!");
        return null;
    }

    protected void init() {
        versionsAdapter = new VersionSpinnerAdapter(
                getContext(),
                android.R.layout.simple_spinner_item,
                VersionRepository.versionList);
        versionsAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        versionSpinner.setAdapter(versionsAdapter);

        LanguageRepository languageRepository = new LanguageRepository(getContext());
        languageAdapter = new LanguageSpinnerAdapter(
                getContext(),
                android.R.layout.simple_spinner_item,
                languageRepository.languageList
        );
        languageAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        languageSpinner.setAdapter(languageAdapter);

        projectTitle.setText(projectName);
        languageSpinner.setSelection(languageAdapter.getPosition(language));
        versionSpinner.setSelection(versionsAdapter.getPosition(version));

        if(isTransforming) {
            button.setEnabled(false);
            progress.setVisibility(View.VISIBLE);
            languageSpinner.setVisibility(View.GONE);
            versionSpinner.setVisibility(View.GONE);
        }

        button.setText(buttonText);
        messageView.setText(messageText);

        allBooksCheckbox.setText(getString(R.string.all_books_check, projectsName));

        button.setOnClickListener(new View.OnClickListener(){
            public void onClick(View v) {
                transform();
            }
        });
    }
}

File: app/src/main/java/bible/translationtools/converter/TransformerFragment.kt
package bible.translationtools.converter

import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.fragment.app.Fragment
import bible.translationtools.converter.databinding.TransformerFragmentBinding
import bible.translationtools.converterlib.ITransformer
import bible.translationtools.converterlib.Project
import bible.translationtools.converterlib.Transformer
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class TransformerFragment : Fragment() {

    @Inject lateinit var directoryProvider: DirectoryProvider
    @Inject lateinit var langRepo: LanguageRepository
    @Inject lateinit var versionRepo: VersionRepository
    @Inject lateinit var transform: Transform

    private lateinit var projectName: String
    private lateinit var projectsName: String // language and version
    private lateinit var language: Language
    private lateinit var version: Version
    private lateinit var book: String

    private var isTransforming = false
    private var messageText = ""
    private var buttonText = ""

    private var newLanguage: Language? = null
    private var newVersion: Version? = null

    private lateinit var versionsAdapter: ArrayAdapter<Version>
    private lateinit var languageAdapter: ArrayAdapter<Language>

    private var transformer: ITransformer? = null

    private var _binding: TransformerFragmentBinding? = null
    private val binding get() = _binding!!

    private val uiScope = CoroutineScope(Dispatchers.Main)

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        super.onCreateView(inflater, container, savedInstanceState)
        _binding = TransformerFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        buttonText = getString(R.string.transform) // Default value
        binding.languages.setTitle(getString(R.string.select_language))

        try {
            readBundle(requireArguments())
            init()
        } catch (e: Exception) {
            finishWithError(e.message ?: "Invalid parameters")
        }
    }

    private fun transform() {
        newLanguage = binding.languages.selectedItem as Language?
        newVersion = binding.versions.selectedItem as Version?

        newLanguage = if (newLanguage != language) newLanguage else null
        newVersion = if (newVersion != version) newVersion else null

        val newLanguageSlug = newLanguage?.slug
        val newVersionSlug = newVersion?.slug

        val transformAll = binding.allBooksCheckbox.isChecked

        try {
            transformationStarted()

            transformer = Transformer(
                directoryProvider.workspaceDir.absolutePath,
                language.slug,
                version.slug,
                (if (!transformAll) book else null),
                newLanguageSlug,
                null,
                newVersionSlug,
                false
            )

            val handler = Handler(Looper.getMainLooper())
            uiScope.launch(Dispatchers.IO) {
                val result = transform(transformer!!)
                handler.post { transformationDone(result.count) }
            }
        } catch (e: Exception) {
            Toast.makeText(context, e.message, Toast.LENGTH_LONG).show()
            e.printStackTrace()
        }
    }

    private fun transformationStarted() {
        isTransforming = true
        messageText = ""
        buttonText = getString(R.string.processing)

        binding.transform.isEnabled = false
        binding.transform.text = buttonText
        binding.languages.visibility = View.GONE
        binding.versions.visibility = View.GONE
        binding.allBooksCheckbox.visibility = View.GONE
        binding.progressBar.visibility = View.VISIBLE
        binding.messageView.text = messageText
    }

    private fun transformationDone(result: Int) {
        isTransforming = false
        buttonText = getString(R.string.transform)
        transformer?.setDateTimeDir()
        binding.transform.isEnabled = true
        binding.transform.text = buttonText
        binding.progressBar.visibility = View.GONE
        binding.languages.visibility = View.VISIBLE
        binding.versions.visibility = View.VISIBLE

        if (result >= 0) {
            messageText = getString(R.string.transformation_complete, result)
            binding.messageView.text = messageText

            language = newLanguage ?: language
            version = newVersion ?: version
            projectName = String.format("%s | %s | %s", language.slug, version.slug, book)
            projectsName = String.format("%s | %s", language.slug, version.slug)

            binding.titleView.text = projectName
            binding.languages.setSelection(languageAdapter.getPosition(language))
            binding.versions.setSelection(versionsAdapter.getPosition(version))
            binding.allBooksCheckbox.text = getString(R.string.all_books_check, projectsName)

            requireArguments().putString("projectName", projectName)
            requireArguments().putString("language", language.slug)
            requireArguments().putString("version", version.slug)
        } else {
            messageText = getString(R.string.error_occurred)
            binding.messageView.text = messageText
        }
        binding.messageView.setTextColor(Color.BLACK)
        println("Finished!")
    }

    private fun init() {
        versionsAdapter = VersionSpinnerAdapter(
            requireContext(),
            android.R.layout.simple_spinner_item,
            versionRepo.versionList
        )
        versionsAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.versions.adapter = versionsAdapter

        languageAdapter = LanguageSpinnerAdapter(
            requireContext(),
            android.R.layout.simple_spinner_item,
            langRepo.languageList
        )
        languageAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.languages.adapter = languageAdapter

        binding.titleView.text = projectName
        binding.languages.setSelection(languageAdapter.getPosition(language))
        binding.versions.setSelection(versionsAdapter.getPosition(version))

        if (isTransforming) {
            binding.transform.isEnabled = false
            binding.progressBar.visibility = View.VISIBLE
            binding.languages.visibility = View.GONE
            binding.versions.visibility = View.GONE
        }

        binding.transform.text = buttonText
        binding.messageView.text = messageText

        binding.allBooksCheckbox.text = getString(R.string.all_books_check, projectsName)

        binding.transform.setOnClickListener { transform() }
    }

    private fun readBundle(bundle: Bundle) {
        projectName = bundle.getString("projectName") ?: throw Exception("Project name not defined")

        val langArg = bundle.getString("language") ?: throw Exception("Language not defined")
        val lang = langRepo.getLang(langArg)

        if (lang == null) {
            throw Exception("Language not found")
        }

        language = Language(lang.slug, lang.name, lang.angName)

        val verArg = bundle.getString("version") ?: throw Exception("Version not defined")
        val ver = versionRepo.getVersion(verArg)

        if (ver == null) {
            throw Exception("Version not found")
        }

        version = Version(ver.slug, ver.name)
        book = bundle.getString("book") ?: throw Exception("Book not defined")

        projectsName = String.format("%s | %s", langArg, verArg)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun finishWithError(error: String) {
        val bundle = Bundle()
        bundle.putString(TRANSFORMER_ERROR_KEY, error)
        parentFragmentManager.setFragmentResult(TRANSFORMER_ID, bundle)
        parentFragmentManager.popBackStack()
    }

    companion object {
        const val TRANSFORMER_ID = "transformer"
        const val TRANSFORMER_ERROR_KEY = "error"

        fun newInstance(project: Project): TransformerFragment {
            val bundle = Bundle()
            bundle.putString("projectName", project.toString())
            bundle.putString("language", project.language)
            bundle.putString("version", project.version)
            bundle.putString("book", project.book)

            val fragment = TransformerFragment()
            fragment.setArguments(bundle)
            return fragment
        }
    }
}
